### ISSUE TITLE
 How to use OAuth2 tokens with UserRequest endpoints?

### ISSUE THREAD

**Issue by chrissound** [discord-haskell/discord-haskell#213]
The library includes `UserRequest` data constructors that seem designed for OAuth2 endpoints (like `GetCurrentUserGuilds` which requires the "guilds" OAuth2 scope), but I can't get them to work with OAuth2 tokens (specifically with OAuth2 access tokens).

Example code:
```haskell
getUserGuilds :: Text -> IO [PartialGuild]
getUserGuilds token = do
    logChan <- newChan
    (handle, tid) <- startRestThread (Auth token) logChan
    result <- writeRestCall handle GetCurrentUserGuilds
    case result of
        Left err -> do
            putStrLn $ "Failed to get guilds: " ++ show err
            return []
        Right guilds -> return guilds
```

Which results in an `RestCallInternalErrorCode 401 "Unauthorized" "{\"message\": \"401: Unauthorized\", \"code\": 0}"` error.

This results in a 401 Unauthorized error, even with a valid OAuth2 token (that I verified working with curl):

```bash
curl -H "Authorization: Bearer $TOKEN" https://discord.com/api/users/@me/guilds
# ^ Works correctly and returns guilds
```

**Comment by yutotakano** [comment]
Hi! The `Auth` type content is basically just passed as-is to the Authorization HTTP header — sorry that it's unintuitive! It can be found in `src/Discord/Internal/Rest/Prelude.hs` under the `authHeader` function.

Basically, just changing your token string to include the "Bearer " part should make it work. Let us know if this doesn't fix it!

**Comment by chrissound** [comment]
Ah sorry I forgot to mention I did try that but got the same error! I can prepare a full minimal example if it helps.

**Comment by yutotakano** [comment]
Hmm, I'm occupied for the next few hours to do a full depth investigation unfortunately. If no one else on the maintainer team gets to it by the time I'm free, I'll find what's wrong!

**Comment by chrissound** [comment]
I've managed to use the mitm proxy to figure out that it's sending the following header:

`Authorization:	Bot Bearer *********************`

It seems to be prepending the `Bot ` infront which I think is the culprit here. 

I used the following changes for it to use the mitm proxy (if it's helpful for in future) (`sudo mitmweb --mode reverse:https://discord.com  -p 443`):
```
diff --git a/src/Discord/Internal/Rest/Prelude.hs b/src/Discord/Internal/Rest/Prelude.hs
index c704076..22c0c5a 100644
--- a/src/Discord/Internal/Rest/Prelude.hs
+++ b/src/Discord/Internal/Rest/Prelude.hs
@@ -12,6 +12,11 @@ import Control.Monad.IO.Class (MonadIO, liftIO)
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as TE
 
+-- Add these imports at the top
+import Network.Connection (TLSSettings(TLSSettingsSimple))
+import Network.HTTP.Client (newManager)
+import Network.HTTP.Client.TLS (mkManagerSettings)
+
 
 import qualified Network.HTTP.Req as R
 
@@ -23,7 +28,7 @@ apiVersion = "10"
 
 -- | The base url (Req) for API requests
 baseUrl :: R.Url 'R.Https
-baseUrl = R.https "discord.com" R./: "api" R./: apiVersion'
+baseUrl = R.https "localhost" R./: "api" R./: apiVersion'
   where apiVersion' = "v" <> apiVersion
 
 -- | Discord requires HTTP headers for authentication.
@@ -62,8 +67,19 @@ class Request a where
 newtype RestIO a = RestIO { restIOtoIO :: IO a }
   deriving (Functor, Applicative, Monad, MonadIO)
 
+--instance R.MonadHttp RestIO where
+  ---- | Throw actual exceptions
+  --handleHttpException = liftIO . throwIO
+  ---- | Don't throw exceptions on http error codes like 404
+  --getHttpConfig = pure $ R.defaultHttpConfig { R.httpConfigCheckResponse = \_ _ _ -> Nothing }
+
 instance R.MonadHttp RestIO where
   -- | Throw actual exceptions
   handleHttpException = liftIO . throwIO
   -- | Don't throw exceptions on http error codes like 404
-  getHttpConfig = pure $ R.defaultHttpConfig { R.httpConfigCheckResponse = \_ _ _ -> Nothing }
+  getHttpConfig = do
+    manager <- liftIO $ newManager $ mkManagerSettings (TLSSettingsSimple True False False) Nothing
+    pure $ R.defaultHttpConfig 
+      { R.httpConfigCheckResponse = \_ _ _ -> Nothing
+      , R.httpConfigAltManager = Just manager
+      }
```

**Comment by chrissound** [comment]
This fix works for me, perhaps you @yutotakano or someone more familiar with the token handling can have a look and add this in? Thanks for your help! 

src/Discord/Internal/Types/Prelude.hs:
```
-- | Get the raw token formatted for use with the websocket gateway
authToken :: Auth -> T.Text
authToken (Auth tok) = let token = T.strip tok
                           botPrefix = if "Bot " `T.isPrefixOf` token || "Bearer " `T.isPrefixOf` token
                                then ""
                                else "Bot "
                       in botPrefix <> token
```

**Comment by L0neGamer** [comment]
Interesting; why doesn't Bot Bearer work for you? Are you using this with a non-bot account somehow?

If this is all that's sufficient feel free to make a PR!

**Comment by chrissound** [comment]
No it's a bot account/application. It responds with a 401 if I do 

```
curl -s \
  'https://discord.com/api/users/@me/guilds' \
  -H "Authorization: Bot Bearer $ACCESS_TOKEN"
```

So I don't think that `Bot Bearer ...` is valid?

**Comment by yutotakano** [comment]
> src/Discord/Internal/Types/Prelude.hs:

Ahhhh, you're right. Missed the existence of that function, I forgot Bots use a different prefix than normal OAuth tokens and assumed all of it used `Bearer`.

To clarify, you're using an OAuth2 workflow to get the user token, correct? i.e. when you do /users/@me via cURL you get *yourself* and not *the bot* as the user.

To explain, in Discord, you can have either an OAuth2 user token generated from doing the OAuth2 workflow (e.g. "login with discord") or a bot token. The user token is for performing actions on behalf of the user, while the bot token is for doing actions as the bot. The OAuth2-generated token needs a "Bearer 123123123" format, the bot token needs a "Bot 123123123" format. The former cannot use the websocket gateway (only limited REST endpoints as specified by the user-accepted scope), while the latter has full capability of the entire Discord API. 

The latter is the intended use of this library, but we can definitely add your fix, which will allow using OAuth2 user tokens too. However, you'll have to bear in mind that it will only work for the HTTP endpoints allowed by your scope, and you won't be able to use the bot gateway with that token.

**Comment by chrissound** [comment]
Yes I'm doing an OAuth2 workflow in this particular case, thanks for the explanation!

However on that point although it is mostly clear to me now about this authentication functionality and limitations, it wasn't clear at all when I first started reading the haddocks or source code (in fact the only clue was these `UserRequest` types, so perhaps this can ideally be mentioned somewhere though I'm not sure what a good place for this is? At least this thread can now be found via a web search.

It's great though that this library provides support for these 'user' endpoints - my earlier function `getUserGuilds :: Text -> IO [PartialGuild]` worked as expected.

**Comment by yutotakano** [comment]
I've opened a PR to add your fix! While that's the implementation fix, I'll keep this issue open until we have more visible documentation somewhere about the two differing ways to use this library, as you say. 

I think we had an unconscious/implicit assumption that those using this library would have some familiarity with the Discord API already. It'd be better if we had more beginner-friendly documentation and that's definitely a todo for this library -- if you find other places of ambiguity or have other questions, please don't hesitate to raise more issues!

**Comment by chrissound** [comment]
Sounds great, thanks! I actually find it easier to read this library to understand the discord API as opposed to the other way round, so definitely the docs here would be valuable I'd imagine.

**Comment by L0neGamer** [comment]
That's quite entertaining for me because for some of the issues we've been having recently I've been looking at other librarys' source to figure out what's been expected; it hasn't been overly fruitful, but it's been at least a little useful
